### PR TITLE
Abstracting prompting transformer for use in L2P and S-Prompt

### PR DESCRIPTION
**Pull request by prabhuteja12**
The previous version of L2P had the "prompting" mechanism tightly coupled. This PR separates the Prompting and the prompt selection strategy. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/renate/benchmark/models/l2p.py [hunk 1]
```diff
@@ -108,6 +108,123 @@ def forward(self, x: torch.Tensor, manual_prompt_indices: Optional[torch.LongTen
...
+        """
+        if prompt is None:
+            return (
+                self.transformer.get_features(x)
+                if self._is_text_transformer
+                else self.transformer.get_features(x, cls_feat=cls_feat)
+            )
+            # text transformers dont support cls_feat.
+        else:
+            if self._is_text_transformer:
```
Comment (wistuba) [inline src/renate/benchmark/models/l2p.py:197]: we can flatten the if else(if else) part to if, elif, else

File: src/renate/benchmark/models/l2p.py [hunk 2]
```diff
@@ -166,34 +283,22 @@ def __init__(
...
+            mlp_dim=mlp_dim,
+            dropout=dropout,
+            attention_dropout=attention_dropout,
+            num_outputs=num_outputs,
+            add_icarl_class_means=add_icarl_class_means,
+            prediction_strategy=prediction_strategy,
+        )
         prompter = PromptPool(
-            embedding_dim=transformer._embedding_size,
+            embedding_dim=transformer.transformer._embedding_size,
```
Comment (wistuba) [inline src/renate/benchmark/models/l2p.py:301]: in the following, we access a lot of protected attributes of the transformer. do we want to keep it that way or rather make them public?

File: src/renate/benchmark/models/l2p.py (hunk 1 above)
Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py:197]: Fixed in d8ef1c81da1e7e1e714a7ef96ab8beb775f0bad4

File: src/renate/benchmark/models/l2p.py [hunk 3]
```diff
@@ -108,6 +108,123 @@ def forward(self, x: torch.Tensor, manual_prompt_indices: Optional[torch.LongTen
         return selected_prompts, loss_value
 
 
+class PromptedTransformer(nn.Module):
+    """This generic module is the basic prompted transformer. It takes in a model string and creates
+    the appropriate transformer. If not prompted, it returns features, and if prompted, it returns
+    the full feature using those prompts and the input image/text.
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py:114]: Fixed in 3896135f37f3e71f424265c36d75025819986c8b.

File: src/renate/benchmark/models/l2p.py (hunk 2 above)
Comment (prabhuteja12) [reply src/renate/benchmark/models/l2p.py:301]: Fixed in 3896135f37f3e71f424265c36d75025819986c8b

**Review by wistuba — APPROVED** [review]

**Pull request merged**